File: Project Enhancements/Software Design & Engineering/Brea_SoftDesignEngineeringEnhancement/Source/src/TaskService.cpp
#include "TaskService.h"
#include <stdexcept>

void TaskService::addTask(const std::string& id, const std::string& name, const std::string& description) {
    //validate ID
    if (id.empty() || id.length() > 10) {
        throw std::invalid_argument("Invalid ID");
    }

    //add new task to map
    tasks.emplace(id, Task(id, name, description));
}

//finds and removes a task from the map via its id
void TaskService::deleteTask(const std::string& id) {
    tasks.erase(id);
}

//finds and updates a task from the map via its id and given paramters
void TaskService::updateTask(const std::string& id, const std::string& name, const std::string& description) {
    auto it = tasks.find(id);
    if (it != tasks.end()) {
        it->second.setName(name);
        it->second.setDescription(description);
    }
}

//finds and updates a task's name from the map via its id and given parameters
void TaskService::updateTaskName(const std::string& id, const std::string& name) {
    tasks.at(id).setName(name);
}

//finds and updates a task's description from the map via its id and given parameters
void TaskService::updateTaskDescription(const std::string& id, const std::string& description) {
    tasks.at(id).setDescription(description);
}

//finds and returns a task from the map via its id
Task TaskService::getTask(const std::string& id) {
    return tasks.at(id);
}

File: Project Enhancements/Brea Software Design and Engineering Enhancement/BreaCS320Proj1/src/main/TaskService.java
package main;

import java.util.HashMap;

public class TaskService {
	private HashMap<String, Task> service = new HashMap<String, Task>();
	
	/**
	 * @param Add Task to Task Service using ID, name, and description
	 */
	public void AddTask(String ID, String name, String desc) {
		//check ID before setting it
		if (ID == null || ID.length() > 10 || ID.isEmpty()) {
			throw new IllegalArgumentException("Invalid ID");
		}
		
		//create new Task Object and input into HashMap service
		service.put(ID, new Task(ID, name, desc));
	}
	
	/**
	 * @param delete Task based on given ID
	 * @return true if removed
	 */
	public void DeleteTask(String ID) {
		//if ID is found, removes task
		service.remove(ID);
	}
	
	/**
	 * @param update name and description using ID
	 */
	public void UpdateTask(String ID, String name, String desc) {
		service.get(ID).setName(name);
		service.get(ID).setDescription(desc);
	}
	
	/**
	 * @param update name using ID
	 */
	public void UpdateTaskName(String ID, String name) {
		service.get(ID).setName(name);
	}
	
	/**
	 * @param update description using ID
	 */
	public void UpdateTaskDescription(String ID, String desc) {
		service.get(ID).setDescription(desc);
	}
	
	/**
	 * @return Task based on given ID
	 */
	public Task getTask(String ID) {
		return service.get(ID);
	}
}
